tests/robustness/watch.go

```diff
@@ -45,35 +46,59 @@ func collectClusterWatchEvents(ctx context.Context, t *testing.T, lg *zap.Logger
 		if err != nil {
 			t.Fatal(err)
 		}
-
+		memberChan := make(chan int64, 1)
+		memberMaxRevisionChans[i] = memberChan
 		wg.Add(1)
 		go func(i int, c *clientv3.Client) {
 			defer wg.Done()
 			defer c.Close()
-			responses := watchMember(ctx, lg, c)
+			responses := watchMember(lg, c, memberChan)
 			mux.Lock()
 			memberResponses[i] = responses
 			mux.Unlock()
 		}(i, c)
 	}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		maxRevision := <-maxRevisionChan
+		for _, memberChan := range memberMaxRevisionChans {
+			memberChan <- maxRevision
+		}
+	}()
 	wg.Wait()
 	return memberResponses
 }
 
-func watchMember(ctx context.Context, lg *zap.Logger, c *clientv3.Client) (resps []watchResponse) {
+func watchMember(lg *zap.Logger, c *clientv3.Client, maxRevisionChan <-chan int64) (resps []watchResponse) {
 	var lastRevision int64 = 0
+	var maxRevision int64 = 0
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	watch := c.Watch(ctx, "", clientv3.WithPrefix(), clientv3.WithRev(lastRevision+1), clientv3.WithProgressNotify())
 	for {
 		select {
 		case <-ctx.Done():
 			return resps
-		default:
-		}
-		for resp := range c.Watch(ctx, "", clientv3.WithPrefix(), clientv3.WithRev(lastRevision+1), clientv3.WithProgressNotify()) {
+		case maxRevision = <-maxRevisionChan:
+			if lastRevision >= maxRevision {
+				cancel()
+			}
+		case resp := <-watch:
 			resps = append(resps, watchResponse{resp, time.Now()})
```

Comment:
Shell we have a case if resp is nil (it is watch stream got terminated on the server or watch stream ctx got cancelled ) ?

Reply:
resp is not a pointer so we will never get an nil. In cases you mentioned I expect resp will hold an error. In such case we report it in logs. Still I don't see reason why not append it. This function should collect all responses so during validation step we can check it.